File: AlgorithmsAndDataStructures/lab07/K.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

const int32_t MAXN = 500000;
vector<vector<int32_t>> graph(MAXN);
vector<pair<int32_t, int32_t>> queries;
int32_t k;

vector<pair<int32_t, int32_t>> euler_tour;
vector<int> tin(MAXN);
vector<int> tout(MAXN);
int32_t counter = 0;

vector<vector<pair<int32_t, int32_t>>> sparse_table;
vector<int32_t> lg;

pair<int32_t, int32_t> min(pair<int32_t, int32_t> a, pair<int32_t, int32_t> b) {
  if (a.first <= b.first) {
    return a;
  }
  return b;
}

void calc_log(int32_t n) {
  lg.resize(n + 1, 0);
  for (int i = 2; i <= n; ++i) {
    lg[i] = lg[i / 2] + 1;
  }
}

void build_sp() {
  int32_t n = euler_tour.size();
  calc_log(n);
  sparse_table.resize(n, vector<pair<int32_t, int32_t>> (lg[n] + 1));
  for (int32_t i = 0; i < n; ++i) {
    sparse_table[i][0] = euler_tour[i];
  }
  for (int32_t k = 1; k <= lg[n]; ++k) {
    for (int32_t i = 0; i < n - (1 << (k - 1)); ++i) {
      sparse_table[i][k] = min(sparse_table[i][k - 1], sparse_table[i + (1 << (k - 1))][k - 1]);
    }
  }
}

pair<int32_t, int32_t> sp_get(int32_t l, int32_t r) {
  int32_t log = lg[r - l + 1];
  return min(sparse_table[l][log], sparse_table[r - (1 << log) + 1][log]);
}

void dfs(int32_t v = 0, int32_t depth = 0) {
  tin[v] = counter++;
  euler_tour.push_back(make_pair(depth, v));
  for (auto u: graph[v]) {
    dfs(u, depth + 1);
    euler_tour.push_back(make_pair(depth, v));
  }
  tout[v] = counter++;
}

bool is_ancestor(int32_t u, int32_t v) {
  return tin[u] <= tin[v] && tout[u] >= tout[v];
}

int32_t LCA(int32_t u, int32_t v) {
  if (is_ancestor(u, v)) return u;
  if (is_ancestor(v, u)) return v;
  if (tin[u] >= tin[v]) {
    swap(u, v);
  }
  return sp_get(tout[u], tin[v]).second;
}

void in() {
  ios_base::sync_with_stdio(0);
  cin.tie(0);
  cout.tie(0);
  cin >> k;
  for (int i = 0; i < k; i++) {
    string s;
    cin >> s;
    int32_t a, b;
    cin >> a >> b;
    if (s == "ADD") {
      graph[a - 1].push_back(b - 1);
    } else {
      queries.push_back(make_pair(a - 1, b - 1));
    }
  }
}

int main() {
  in();
  dfs();
  build_sp();
  for (auto [u, v] : queries) {
    cout << LCA(u, v) + 1 << "\n";
  }
  return 0;
}

File: AlgorithmsAndDataStructures/lab07/K.kt
import java.io.PrintWriter
import java.util.*

private val INPUT = System.`in`
private val OUTPUT = System.out
private val _reader = INPUT.bufferedReader()
private val _writer = PrintWriter(OUTPUT, false)
private var _tokenizer: StringTokenizer = StringTokenizer("")

private fun read(): String {
    while (_tokenizer.hasMoreTokens().not()) {
        _tokenizer = StringTokenizer(_reader.readLine() ?: return "", " ")
    }
    return _tokenizer.nextToken()
}

private fun readLn() = _reader.readLine()!!
private fun readInt() = read().toInt()
private fun readInts(n: Int) = List(n) { read().toInt() }

private inline fun output(block: PrintWriter.() -> Unit) {
    _writer.apply(block).flush()
}

fun main() {
    val n = 500_000
    val graph = Array(n) { ArrayList<Int>() }
    val queries = ArrayList<Pair<Int, Int>>()
    val k = readInt()
    for (i in 1..k) {
        val cmd = read()
        val (a, b) = readInts(2)
        if (cmd == "ADD") {
            graph[a - 1].add(b - 1)
        } else {
            queries.add(Pair(a - 1, b - 1))
        }
    }
    val tree = TreeK(n, graph)
    tree.initLCA()
    output {
        for ((u, v) in queries) {
            println(tree.LCA(u, v) + 1)
        }
    }
}

class TreeK(
    private val n: Int,
    private val graph: Array<ArrayList<Int>>,
) {
    private val timeIn = IntArray(n)
    private val timeOut = IntArray(n)
    private val depth = IntArray(n)
    private var counter = 0

    private var sparseTable: SparseTableInt? = null
    private var eulerTour: IntArray? = null

    fun initLCA() {
        val eulerTour_ = IntArray(2 * n - 1)
        dfs(0, 0, eulerTour_, 0)
        sparseTable = SparseTableInt(eulerTour_) { a, b -> if (depth[a] <= depth[b]) a else b }
        eulerTour = eulerTour_
    }

    private fun dfs(v: Int, curDepth: Int, list: IntArray, idx: Int): Int {
        depth[v] = curDepth
        timeIn[v] = counter++
        list[idx] = v
        var nextIdx = idx + 1
        for (u in graph[v]) {
            val newIdx = dfs(u, curDepth + 1, list, nextIdx)
            list[newIdx] = v
            nextIdx = newIdx + 1
        }
        timeOut[v] = counter++
        return nextIdx
    }

    fun isAncestor(u: Int, v: Int) = timeIn[u] <= timeIn[v] && timeOut[u] >= timeOut[v]

    fun LCA(u: Int, v: Int): Int {
        if (isAncestor(u, v)) return u
        if (isAncestor(v, u)) return v
        val flag = timeIn[u] < timeIn[v]
        val a = if (flag) u else v
        val b = if (flag) v else u
        return sparseTable!!.get(timeOut[a], timeIn[b])
    }


    class SparseTableInt(
        private val array: IntArray,
        private val f: (Int, Int) -> Int,
    ) {
        private val log = calcLog(array.size)
        private val sparseTable = calcSparseTable(array, log, f)

        fun get(l: Int, r: Int): Int {
            val log = log[r - l + 1]
            return f(sparseTable[l][log], sparseTable[r - (1 shl log) + 1][log])
        }

        companion object {
            private fun calcLog(n: Int): IntArray {
                val log = IntArray(n + 1)
                for (i in 2..n) {
                    log[i] = log[i / 2] + 1
                }
                return log
            }

            private fun calcSparseTable(array: IntArray, log: IntArray, f: (Int, Int) -> Int): Array<IntArray> {
                val n = array.size
                val table = Array(n) { IntArray(log[n] + 1) }
                for ((index, list) in table.withIndex()) {
                    list[0] = array[index]
                }
                for (k in 1..log[n]) {
                    for (index in 0 until n - (1 shl (k - 1))) {
                        table[index][k] = f(
                            table[index][k - 1],
                            table[index + (1 shl (k - 1))][k - 1]
                        )
                    }
                }
                return table
            }
        }
    }
}
